aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu

```diff
@@ -0,0 +1,155 @@
+#include <ATen/ATen.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/InitialTensorOptions.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/SparseCsrTensorImpl.h>
+#include <ATen/SparseCsrTensorUtils.h>
+#include <ATen/SparseTensorUtils.h>
+#include <ATen/WrapDimUtilsMulti.h>
+#include <ATen/native/BinaryOps.h>
+#include <algorithm>
+
+#include <cuda_runtime.h>
+#include <type_traits>
+
+#include <THC/THCTensorMathPointwise.cuh>
+#include <THC/THCThrustAllocator.cuh>
+
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAUtils.h>
+#include <c10/cuda/CUDACachingAllocator.h>
+#include <ATen/native/sparse/cuda/SparseCUDABlas.cuh>
+
+#include <thrust/device_ptr.h>
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/sequence.h>
+
+namespace at {
+namespace native {
+
+using namespace at::sparse_csr;
+// certain utiliy functions are usable from sparse COO.
+using namespace at::sparse;
+
+Tensor& add_out_dense_sparse_csr_cuda(
+    Tensor& output,
+    const Tensor& dense,
+    const SparseCsrTensor& src,
+    const Scalar& alpha) {
+  AT_ASSERT(dense.layout() == kStrided);
+  AT_ASSERT(src.is_sparse_csr());
+  AT_ASSERT(dense.is_cuda());
```

Comment:
nit TORCH_INTERNAL_ASSERT

Reply:
Also, you'll get sent here even for mixed CPU-CUDA inputs, so modulo @wenleix's patch set about code generating device checks (which I don't think is being applied for add), you DO need to test this explicitly in the body. Don't sweat this too much, a lot of code in PyTorch gets it wrong and generated device checks are the correct fix.